### PR TITLE
Add tests for raw template expressions

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/35724

### CODE REVIEW

File: conformance/lang/expressions/raw_template_expr.balt [hunk 1]
```diff
@@ -0,0 +1,690 @@
...
+    io:println(a is object:RawTemplate); // @output true
+
+    int b = 1;
+    object:RawTemplate c = `int value ${b}, object value ${a}`;
+    io:println(c is object:RawTemplate); // @output true
+}
+
+Test-Case: output
+Description: Test raw template expression constructing an object belonging to the abstract RawTemplate object type, with a user-defined subtype of `object:RawTemplate` as the contextually-expected type.
+Labels: array-type, intersection-type, module-type-defn, object:RawTemplate, object-type, raw-template-expr
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:29]: missing `readonly-type`

Comment (lasinicl) [reply conformance/lang/expressions/raw_template_expr.balt:29]: Should we include a label for `object-type-inclusion`?

File: conformance/lang/expressions/raw_template_expr.balt [hunk 2]
```diff
@@ -0,0 +1,690 @@
+Test-Case: output
+Description: Test raw template expression constructing an object belonging to the abstract RawTemplate object type, with var.
+Labels: object:RawTemplate, raw-template-expr, var
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:3]: missing `is-expr` label

File: conformance/lang/expressions/raw_template_expr.balt [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+    io:println(a is object:RawTemplate); // @output true
+
+    int b = 1;
+    var c = `int value ${b}, object value ${a}`;
+    io:println(c is object:RawTemplate); // @output true
+}
+
+Test-Case: output
+Description: Test raw template expression constructing an object belonging to the abstract RawTemplate object type, with `object:RawTemplate` as the contextually-expected type.
+Labels: object:RawTemplate, object-type, raw-template-expr
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:16]: missing `is-expr` label

File: conformance/lang/expressions/raw_template_expr.balt [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+    MyRawTemplate a = `number ${i} and number ${2}`;
+
+    io:println(a is object:RawTemplate); // @output true
+    io:println(a.strings); // @output ["number "," and number ",""]
+    io:println(a.insertions); // @output [1,2]
+}
+
+Test-Case: output
+Description: Test raw template expression constructing an object belonging to the abstract RawTemplate object type, with an in-line user-defined subtype of `object:RawTemplate` as the contextually-expected type.
+Labels: array-type, intersection-type, object:RawTemplate, object-type, raw-template-expr
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:49]: missing `readonly-type`, `field-access-expr`

Comment (lasinicl) [reply conformance/lang/expressions/raw_template_expr.balt:49]: Let's update wherever applicable

File: conformance/lang/expressions/raw_template_expr.balt [hunk 5]
```diff
@@ -0,0 +1,690 @@
...
+
+function init() {
+    int i = 1;
+    MyRawTemplate a = `number ${i} and number ${2}`;
+    object:RawTemplate b = a;
+    b.strings = []; // @panic cannot overwrite the read-only `strings` field
+}
+
+Test-Case: panic
+Description: Test the raw template `insertions` fields being read-only when required by the contextually-expected type.
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:277]: Shall we also add a positive test for this where contextually-expected type doesn't require `readonly` and can update the `insertions` fields?

File: conformance/lang/expressions/raw_template_expr.balt [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+    (any|error)[] b = a.insertions;
+    (any|error)[] c = fn().insertions;
+    io:println(b === c); // @output false
+}
+
+function fn() returns object:RawTemplate {
+    int i = 2;
+    return `foo${1}bar${i}`;
+}
+
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:327]: Do we have tests to cover?

> The type of the insertions field is determined from the contextually expected type or the static type of the expressions in the BacktickString in the same way as with a list-constructor-expr.

File: conformance/lang/expressions/raw_template_expr.balt [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+    int i = 1;
+    MyRawTemplate a = `number ${i} and number ${2}`;
+    object:RawTemplate b = a;
+    b.insertions = []; // @panic cannot overwrite the read-only `insertions` field
+}
+
+Test-Case: output
+Description: Test the value of the `strings` field being constructed only once by the class for the raw template expression.
+Fail-Issue: ballerina-platform/ballerina-lang#36354
+Labels: array-type, object:RawTemplate, raw-template-expr
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:298]: missing `function-call-expr`, `exact-equality`

File: conformance/lang/expressions/raw_template_expr.balt [hunk 8]
```diff
@@ -0,0 +1,690 @@
...
+}
+
+function fn() returns object:RawTemplate {
+    int i = 2;
+    return `foo${1}bar${i}`;
+}
+
+Test-Case: output
+Description: Test the value of the `insertions` field being constructed once for each evaluation of the raw template expression.
+Labels: any, array-type, error-type, object:RawTemplate, raw-template-expr, union-type
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:314]: missing `function-call-expr`, `exact-equality`

File: conformance/lang/expressions/raw_template_expr.balt [hunk 9]
```diff
@@ -0,0 +1,690 @@
...
+}
+
+function fn() returns object:RawTemplate {
+    int i = 2;
+    return `foo${1}bar${i}`;
+}
+
+Test-Case: output
+Description: Test literal expressions as interpolations.
+Labels: any, array-type, boolean-literal, check, error-type, nil-literal, object:RawTemplate, raw-template-expr, string-literal, union-type, value:toBalString
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:330]: missing `array:length`, `foreach-stmt`, `iterable-list`

Comment (lasinicl) [reply conformance/lang/expressions/raw_template_expr.balt:330]: Let's update wherever applicable

File: conformance/lang/expressions/raw_template_expr.balt [hunk 10]
```diff
@@ -0,0 +1,690 @@
...
+    foreach any|error item in insertions {
+        io:println(item); // @output [1,error("oops!")]
+                          // @output {"a":2,"b":error("oops!")}
+                          // @output [{"c":3,"e":error("oops!")}]
+    }
+}
+
+Test-Case: output
+Description: Test template expressions as interpolations.
+Labels: any, array-type, check, error-type, field-access-expr, module-class-defn, new-expr, object-constructor-expr, object:RawTemplate, raw-template-expr, type-cast-expr, union-type
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:386]: missing `member-access-expr`

File: conformance/lang/expressions/raw_template_expr.balt [hunk 11]
```diff
@@ -0,0 +1,690 @@
...
+
+function init() {
+    Record rec = {a: 1, b: 2};
+    object:RawTemplate a = `1 ${rec.a} 2 ${rec?.b} 3 ${rec?.c} 4 ${rec["b"]} 5 ${rec["c"]}`;
+    io:println(a.insertions); // @output [1,2,null,2,null]
+}
+
+Test-Case: output
+Description: Test xml attribute access expressions as interpolations.
+Labels: object:RawTemplate, raw-template-expr, xml, xml-optional-attribute-access-expr, xml-required-attribute-access-expr
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:419]: missing `field-access-expr`

Comment (lasinicl) [reply conformance/lang/expressions/raw_template_expr.balt:419]: same for the next set of tests

File: conformance/lang/expressions/raw_template_expr.balt [hunk 12]
```diff
@@ -0,0 +1,690 @@
...
+
+function init() {
+    Class c = new;
+    object:RawTemplate a = `func ${fn()} method ${c.fn()}`;
+    io:println(a.insertions); // @output [true,123]
+}
+
+Test-Case: output
+Description: Test anonymous function expressions as interpolations.
+Labels: additive-expr, check, error-type, explicit-anonymous-function-expr, function-call-expr, function-type, object:RawTemplate, raw-template-expr, type-cast-expr, union-type
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:470]: `member-access-expr`

File: conformance/lang/expressions/raw_template_expr.balt [hunk 13]
```diff
@@ -0,0 +1,690 @@
...
+
+function init() {
+    int a = 10;
+    object:RawTemplate b = `${a & 2} ${a | 2} ${2 ^ a}`;
+    io:println(b.insertions); // @output [2,10,8]
+}
+
+Test-Case: output
+Description: Test range expressions as interpolations.
+Labels: check, field-access-expr, is-expr, object:Iterable, object:RawTemplate, range-expr, raw-template-expr, var, while-stmt
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:558]: missing `type-cast-expr`, `member-access-expr`, `error-type`, `nil-type`, `optional-type`

File: conformance/lang/expressions/raw_template_expr.balt [hunk 14]
```diff
@@ -0,0 +1,690 @@
...
+
+function init() returns error? {
+    int|error a = 1;
+    object:RawTemplate c = `${check a} ${checkpanic a}`;
+    io:println(c.insertions); // @output [1,1]
+}
+
+Test-Case: output
+Description: Test trap expressions as interpolations.
+Labels: error-type, error-constructor-expr, object:RawTemplate, panic, raw-template-expr, trap-expr, union-type
```
Comment (lasinicl) [inline conformance/lang/expressions/raw_template_expr.balt:652]: missing `function-call-expr`

File: conformance/lang/expressions/raw_template_expr.balt (hunk 6 above)
Comment (MaryamZi) [reply conformance/lang/expressions/raw_template_expr.balt:327]: Tests such as https://github.com/ballerina-platform/ballerina-spec/blob/4547bc6db2cdc2ec7de60f5e1bc6319d834f6fa4/conformance/lang/expressions/raw_template_expr.balt#L166 cover the scenario where there is a contextually-expected type IMO.

Will add a test with var for when there's no contextually-expected type.

**Review by lasinicl — APPROVED** [review]
LGTM.
Verified against the test runner. Need to add `value:cloneReadOnly`, `xml:children`, `object:Iterable` labels to `labels.csv` before merging this.

**Pull request merged**